### PR TITLE
AWS - Application autoscaling

### PR DESCRIPTION
**Pull request by cplee**
This change creates a new base action for application autoscaling. Additionally, this PR includes the following changes:

* Add support for onhour/offhour filters to ECS Service
* Add support for `autoscaling` action to ECS Service
* Create new resource named `ec2-spot-fleet-request`
* Add support for onhour/offhour filters to Spot Fleet Request
* Add support for `autoscaling` action to Spot Fleet Request

### CODE REVIEW

**Comment by kapilt** [comment]
super cool, thanks

**Comment by cplee** [comment]
@kapilt - what's the process/next steps to get this merged?

**Review by kapilt — COMMENTED** [review]
> @kapilt - what's the process/next steps to get this merged?

we need someone to review, its a big pr though, i'll try and make some time this week. i went through a quick read again today, the only stylistic comment would be if we could break apart the base action process method, and if we need a validate method to ensure that save/restore are mutually exclusive configuration (alternatively in jsonschema).

**Comment by cplee** [comment]
Thanks for the feedback @kapilt  - I've made updates to reduce the complexity of the `process()` method

**Comment by kapilt** [comment]
thanks @cplee looks good, only remaining questions i have are on naming. 

- generally actions have a verb. autoscaling in this context generally feels like a noun reference, so was thinking it would be good to rename to `set-autoscaling` 
- the `suspend-scaling` `restore-scaling` feel like they could ditch the `-scaling` suffix as that's implied by the action itself
- i'm wondering about the separate parameters for the restore/save tags, and just calling it `tag`, since those actions are mutually exclusive.
-

**Comment by cplee** [comment]
@kapilt - great feedback!

* good catch on the action name. I went ahead and renamed to `resize` to match the way `asg` actions works. Seem reasonable?
* since i went with `resize` instead of `set-autoscaling` it seems i should keep the `-scaling` suffix to avoid confusion. If you still think i should remove it, i'm happy to do so.
*  the `save-options-tag` and `restore-options-tag` were named to match the `asg.resize` action. OK to keep those?

**Comment by kapilt** [comment]
All sound great re resize and tag, thanks!

File: c7n/actions/autoscaling.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        """ return the tag for the provided resource """
+        raise NotImplementedError
+
+    def get_resource_desired(self, resource):
+        """ return the current desired value for the provided resource """
+        raise NotImplementedError
+
+    def set_resource_tag(self, resource, key, value):
+        """ set the tag for the provided resource """
+        raise NotImplementedError
```
Comment (kapilt) [inline c7n/actions/autoscaling.py:100]: we can process this capability generic for any resource that supports tagging, ie thats basically how auto-tagging user works. that then makes get_resource_id extraneous.

**Review by kapilt — COMMENTED** [review]
a couple of questions inline, primary is to clarify intended usage if suspend and restore are intended to be exclusive usage of a given action instance. i'm happy to work on the other feedback items, if you'd prefer, albeit its likely second half of the week.

File: c7n/actions/autoscaling.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+                if resource_id not in resource_ids:
+                    continue
+
+                if target['ScalableDimension'] != self.scalable_dimension:
+                    continue
+
+                resource = resources_by_id[resource_id]
+
+                if 'suspend-scaling' in self.data and self.data['suspend-scaling']:
```
[the remark this answers is not in the document]
Comment (cplee) [reply c7n/actions/autoscaling.py]: done!

File: c7n/actions/autoscaling.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+            policies:
+              - name: offhours-ecs-off
+                resource: ecs-service
+                filters:
+                  - type: offhour
+                    offhour: 19
+                    default_tz: bst
+                actions:
+                  - type: resize
```
[the remark this answers is not in the document]
Comment (cplee) [reply c7n/actions/autoscaling.py:29]: @kapilt  - so this action actually _does_ update the resource. For example, check out the [SpotFleetRequest](https://github.com/cplee/cloud-custodian/blob/application-autoscaling/c7n/resources/ec2.py#L2263-L2268).

This is to match the way that the ASG autoscaling works. I feel this is a better user experience anyway - to treat min/max/desired as a single action. Our use case is to stop services in off-hours and then restart them. I want the min/max/desired to all update together.

[the remark this answers is not in the document]
Comment (kapilt) [reply c7n/actions/autoscaling.py:29]: that's fair.

File: c7n/actions/autoscaling.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                resource = resources_by_id[resource_id]
+
+                if 'suspend-scaling' in self.data and self.data['suspend-scaling']:
+                    # suspend scaling activities
+                    self.process_suspend_scaling(target)
+                elif 'restore-scaling' in self.data and self.data['restore-scaling']:
+                    # restore scaling activities
+                    self.process_restore_scaling(target)
+
+                if 'restore-options-tag' in self.data:
```
[the remark this answers is not in the document]
Comment (cplee) [reply c7n/actions/autoscaling.py:245]: yes - thanks for catching that. just pushed a fix for it.

File: c7n/actions/autoscaling.py (hunk 1 above)
Comment (cplee) [reply c7n/actions/autoscaling.py:100]: I believe I'll still need the `get_resource_id` for other reasons, but i like the idea of getting rid of `set_resource_tag` and replacing with `UniversalTag`. I've pushed a commit with that implementation.

**Comment by cplee** [comment]
@kapilt - the updates you requested are ready to review. There is still an open question on the action name (`resize` vs `set-scaling-policy` or `set-auto-scale`). Please see my comment and see if that changes your opinions on it.

Thanks again for the time you've spent reviewing this PR!

**Comment by cplee** [comment]
@kapilt - do you have any other feedback for me on this PR or is it just waiting for the right time to merge?

**Comment by cplee** [comment]
@kapilt - is this ready to merge?

**Comment by cplee** [comment]
@kapilt - i'd love to get this merged...anything I should be doing to make progress on this one?

**Review by kapilt — APPROVED** [review]
sorry for the delay, lgtm

**Pull request merged**